### PR TITLE
[Producers] Add synchronized variant  

### PR DESCRIPTION
**Pull request by Herklos**
### Added
- Synchronous Channel
- Synchronous Consumer
- Synchronous Producer

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# OctoBot-Channels [1.3.25](https://github.com/Drakkar-Software/OctoBot-Channels/blob/master/CHANGELOG.md)
+# OctoBot-Channels [1.4.0](https://github.com/Drakkar-Software/OctoBot-Channels/blob/master/CHANGELOG.md)
```
Comment (GuillaumeDSM) [inline README.md:1]: :+1:

File: octobot_channels/consumer.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ class Consumer:
     A consumer also responds to channel events like pause and stop.
     """
 
-    def __init__(self, callback: object, size: int = 0):
+    def __init__(self, callback: object, size: int = 0, priority_level: int = 1):
```
Comment (GuillaumeDSM) [inline octobot_channels/consumer.py]: Should we use DEFAULT_PRIORITY_LEVEL here ?

File: octobot_channels/consumer.py [hunk 3]
```diff
@@ -58,9 +62,9 @@ async def consume(self):
                 self.logger.debug("Cancelled task")
             except Exception as consume_exception:  # pylint: disable=broad-except
                 self.logger.exception(
-                    consume_exception,
-                    True,
-                    f"Exception when calling callback on {self}: {consume_exception}",
+                    exception=consume_exception,
+                    publish_error_if_necessary=True,
+                    error_message=f"Exception when calling callback on {self}: {consume_exception}",
```
Comment (GuillaumeDSM) [inline octobot_channels/consumer.py:74]: 👍

File: octobot_channels/producer.py [hunk 4]
```diff
@@ -114,6 +116,19 @@ async def wait_for_processing(self) -> None:
...
+        Empties the queue synchronously for each consumers
+        :param priority_level: the consumer minimal priority level
+        """
+        for consumer in [
+            consumer
+            for consumer in self.channel.get_consumers()
+            if consumer.priority_level <= priority_level
+        ]:
+            while not consumer.queue.empty():
+                await consumer.perform(await consumer.queue.get())
```
Comment (GuillaumeDSM) [inline octobot_channels/producer.py:130]: 👍

File: tests/test_synchronized.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    test_consumer = await synchronized_channel.new_consumer(callback)
+
+    producer = TestSynchronizedProducer(get_chan(TEST_SYNCHRONIZED_CHANNEL))
+    await producer.run()
+
+    with patch.object(test_consumer, 'callback', new=AsyncMock()) as mocked_test_consumer_callback:
+        await producer.send({})
+        await mock_was_not_called(mocked_test_consumer_callback)
+        await producer.synchronized_perform_consumers_queue(1)
+        await mock_was_called_once(mocked_test_consumer_callback)
```
Comment (GuillaumeDSM) [inline tests/test_synchronized.py:64]: 👍

File: tests/test_synchronized.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    test_consumer_3_1 = await synchronized_channel.new_consumer(callback, priority_level=3)
+
+    producer = TestSynchronizedProducer(get_chan(TEST_SYNCHRONIZED_CHANNEL))
+    await producer.run()
+
+    with patch.object(test_consumer_1_1, 'callback', new=AsyncMock()) as mocked_test_consumer_1_1_callback:
+        with patch.object(test_consumer_1_2, 'callback', new=AsyncMock()) as mocked_test_consumer_1_2_callback:
+            with patch.object(test_consumer_2_1, 'callback', new=AsyncMock()) as mocked_test_consumer_2_1_callback:
+                with patch.object(test_consumer_2_2, 'callback', new=AsyncMock()) as mocked_test_consumer_2_2_callback:
+                    with patch.object(test_consumer_3_1, 'callback', new=AsyncMock()) as mocked_test_consumer_3_1_callback:
```
Comment (GuillaumeDSM) [inline tests/test_synchronized.py]: you can chain multiple context managers with a "," 😃

File: tests/test_synchronized.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                        assert not producer.is_consumers_queue_empty(2)
+                        await producer.synchronized_perform_consumers_queue(3)
+                        await mock_was_called_once(mocked_test_consumer_1_1_callback)
+                        await mock_was_called_once(mocked_test_consumer_1_2_callback)
+                        await mock_was_called_once(mocked_test_consumer_2_1_callback)
+                        await mock_was_called_once(mocked_test_consumer_2_2_callback)
+                        await mock_was_called_once(mocked_test_consumer_3_1_callback)
+                        assert producer.is_consumers_queue_empty(1)
+                        assert producer.is_consumers_queue_empty(2)
+                        assert producer.is_consumers_queue_empty(3)
```
Comment (GuillaumeDSM) [inline tests/test_synchronized.py]: 💯

File: octobot_channels/consumer.py (hunk 2 above)
Comment (Herklos) [reply octobot_channels/consumer.py]: right :+1:

File: tests/test_synchronized.py (hunk 6 above)
Comment (Herklos) [reply tests/test_synchronized.py]: Is it possible ? Great !

File: octobot_channels/util/channel_creator.py [hunk 8]
```diff
@@ -19,31 +19,34 @@
 
 
 async def create_all_subclasses_channel(
-    channel_class: ClassVar, set_chan_method: Callable, **kwargs: dict
+    channel_class: ClassVar, set_chan_method: Callable, is_synchronized: bool = False, **kwargs: dict
```
Comment (GuillaumeDSM) [inline octobot_channels/util/channel_creator.py]: 👍

**Review by GuillaumeDSM — APPROVED** [review]
🎉

**Pull request merged**